src/ZkFold/Base/Protocol/Protostar/ArithmeticCircuit.hs

```diff
@@ -0,0 +1,135 @@
+{-# LANGUAGE AllowAmbiguousTypes  #-}
+{-# LANGUAGE DeriveAnyClass       #-}
+{-# LANGUAGE TypeOperators        #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+module ZkFold.Base.Protocol.Protostar.ArithmeticCircuit where
+
+
+import           Control.DeepSeq                                     (NFData)
+import           Data.List                                           (foldl')
+import           Data.Map.Strict                                     (Map)
+import qualified Data.Map.Strict                                     as M
+import           GHC.Generics                                        (Generic)
+import           Prelude                                             (fmap, ($), (.), (<$>), (==))
+import qualified Prelude                                             as P
+
+import           ZkFold.Base.Algebra.Basic.Class
+import           ZkFold.Base.Algebra.Basic.Field
+import           ZkFold.Base.Algebra.Basic.Number
+import qualified ZkFold.Base.Algebra.Polynomials.Multivariate        as PM
+import           ZkFold.Base.Algebra.Polynomials.Multivariate
+import qualified ZkFold.Base.Algebra.Polynomials.Univariate          as PU
+import qualified ZkFold.Base.Data.Vector                             as V
+import           ZkFold.Base.Data.Vector                             (Vector)
+import qualified ZkFold.Base.Protocol.Protostar.SpecialSound         as SPS
+import           ZkFold.Symbolic.Compiler
+import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.Internal
+import           ZkFold.Symbolic.Data.Combinators                    (Iso (..))
+
+{--
+
+1. Compress verification checks (Section 3.5; )
+2. Commit (Section 3.2; ZkFold.Base.Protocol.ARK.Protostar.CommitOpen)
+3. Fiat-Shamir transform (Section 3.3; ZkFold.Base.Protocol.ARK.Protostar.FiatShamir)
+   A technique for taking an interactive proof of knowledge and creating a digital signature based on it.
+   This way, some fact (for example, knowledge of a certain secret number) can be publicly proven without revealing underlying information.
+4. Accumulation scheme (Section 3.4; ZkFold.Base.Protocol.ARK.Protostar.AccumulatorScheme)
+5. Obtain the IVC scheme (Theorem 1 from “Proof-Carrying Data Without Succinct Arguments”; )
+
+--}
+
+
+-- | A type for recurcive computations.
+-- @circuit@ is an Arithmetic circuit with @n@ inputs and @n@ outputs applied to itself (i.e. outputs are fed as inputs at the next iteration) @iterations@ times.
+--
+data RecursiveCircuit n a
+    = RecursiveCircuit
+        { iterations :: Natural
+        , circuit    :: ArithmeticCircuit a (Vector n) (Vector n)
+        } deriving (Generic, NFData)
+
+instance (Ring f, P.Eq f) => Iso (PU.Poly f) f where
```

Comment:
Is this really an isomorphism? Let $p(x) = x+1$, then `from (from p) = 1`

Reply:
No, of course it isn't. I was only trying to make it work and used an available class. I'll create another class

Reply:
Is typeclass needed for this?

Reply:
The problem is, `algebraicMap` from `SpecialSoundProtocol` is polymorphic. Depending on the parameter `f`, it can either construct a system of equations if `f` is a polynomial or just compute a vector if `f` is a field element. This behaviour is required in the algorithm in `AccumulatorScheme`. Compiler doesn't like that `algebraicMap` is called with different type parameters, that's why I gave it a hint that the types can be converted into each other. Let me think if I can do it more nicely.